r2/r2/lib/providers/search/solrsearch.py

```diff
@@ -0,0 +1,742 @@
+                time.sleep(x)
+                continue
+            else:
+                break
+        else:
+            raise err
+        last_update = chunk[-1]
+        g.cache.set(cache_key, last_update._fullname)
+        time.sleep(sleeptime)
+
+
+rebuild_subreddit_index = functools.partial(rebuild_link_index,
+                                            cls=Subreddit,
+                                            uploader=SolrSubredditUploader,
+                                            estimate=200000,
+                                            chunk_size=1000)
+
+
+def test_run_link(start_link, count=1000):
+    '''Inject `count` number of links, starting with `start_link`'''
+    if isinstance(start_link, basestring):
+        start_link = int(start_link, 36)
+    links = Link._byID(range(start_link - count, start_link), data=True,
+                       return_dict=False)
+    uploader = SolrLinkUploader(things=links)
+    return uploader.inject()
+
+
+def test_run_srs(*sr_names):
+    '''Inject Subreddits by name into the index'''
+    srs = Subreddit._by_name(sr_names).values()
+    uploader = SolrSubredditUploader(things=srs)
+    return uploader.inject()
+
+
+def translate_raw_sort(sort):
+    return SORTS_DICT.get(sort, 'score desc') 
+
+
+class SolrSearchProvider(SearchProvider):
+    '''Provider implementation: wrap it all up as a SearchProvider
+    
+    example config:
+
+    # endpoint for link search e.g. http://localhost:8983
+    solr_search_host = is-forums-sf-04v
+    # endpoint for link upload
+    solr_doc_host = is-forums-sf-04v
+    # endpoint for subreddit search
+    solr_subreddit_search_host = is-forums-sf-04v
+    # endpoint for subreddit upload
+    solr_subreddit_doc_host = is-forums-sf-04v
+    # solr port
+    solr_port = 8983
+    # solr core name
+    solr_core = reddit1
+    # default batch size is 500
+    # limit is hard-coded to 1000
+    solr_min_batch = 1
+    
+    '''
+  
+    config = {
+        ConfigValue.int: [
+            "solr_port",
+            "solr_min_batch",
+        ],
+        ConfigValue.str: [
+            "solr_search_host",
+            "solr_doc_host",
+            "solr_subreddit_search_host",
+            "solr_subreddit_doc_host",
+            "solr_core",
+        ],
+    }    
+
+    InvalidQuery = (InvalidQuery,)
+    SearchException = (SearchHTTPError,)
+
+    SearchQuery = LinkSearchQuery
+
+    SubredditSearchQuery = SolrSubredditSearchQuery
+    
+    sorts = LinkSearchQuery.sorts_menu_mapping
+
+    def run_changed(drain=False, min_size=int(getattr(g, 'solr_min_batch', 500)), limit=1000, sleep_time=10, 
+            use_safe_get=False, verbose=False):
+        '''Run by `cron` (through `paster run`) on a schedule to send Things to Solr
+        '''
+        if use_safe_get:
+            SolrSearchUploader.use_safe_get = True
+        amqp.handle_items('cloudsearch_changes', _run_changed, min_size=min_size,
```

Comment:
Just pointing out that this is one that should NOT change (so you did the right thing by leaving it). An unfortunate name choice on my part; perhaps I'll look into renaming it across the board later.